modules/react/common/lib/utils/components.ts

```diff
@@ -148,10 +146,11 @@ interface RefForwardingComponent<T, P = {}> {
  * return type is `Component<element, Props>` which looks like `Component<'div', Props>` which is a
  * clean interface that tells you the default element that is used.
  */
-export const createComponent = <T extends React.ElementType | undefined = undefined>(as?: T) => <
-  P,
-  SubComponents = {}
->({
+export const createComponent = <
+  E extends keyof JSX.IntrinsicElements | React.ComponentType | undefined = undefined
```

Comment:
I extracted out `React.ElementType` to the component parts to avoid confusion. The type is the same though.